README.md
node-coap
=====

![Build Status](https://github.com/mcollina/node-coap/workflows/Build%20Status/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/mcollina/node-coap/badge.svg?branch=master)](https://coveralls.io/github/mcollina/node-coap?branch=master)
[![gitter](https://badges.gitter.im/mcollina/node-coap.png)](https://gitter.im/mcollina/node-coap)

__node-coap__ is a client and server library for CoAP modeled after the `http` module.

  * <a href="#intro">Introduction</a>
  * <a href="#install">Installation</a>
  * <a href="#basic">Basic Example</a>
  * <a href="#proxy">Proxy features</a>
  * <a href="#api">API</a>
  * <a href="#contributing">Contributing</a>
  * <a href="#license">License &amp; copyright</a>

[![NPM](https://nodei.co/npm/coap.png)](https://nodei.co/npm/coap/)

[![NPM](https://nodei.co/npm-dl/coap.png)](https://nodei.co/npm/coap/)

<a name="intro"></a>
## Introduction

What is CoAP?
> Constrained Application Protocol (CoAP) is a software protocol
intended to be used in very simple electronics devices that allows them
to communicate interactively over the Internet. -  Wikipedia

This library follows:
* [RFC 7252](https://datatracker.ietf.org/doc/html/rfc7252) for CoAP,
* [RFC 7641](https://datatracker.ietf.org/doc/html/rfc7641)
  for the observe specification,
* [RFC 7959](https://datatracker.ietf.org/doc/html/rfc7959) for
  the blockwise specification,
* [RFC 8132](https://datatracker.ietf.org/doc/html/rfc8132) for the PATCH, FETCH, and iPATCH methods.

It does not parse the protocol but it use
[CoAP-packet](http://github.com/mcollina/coap-packet) instead.

If you need a command line interface for CoAP, check out
[coap-cli](http://github.com/mcollina/coap-cli).

**node-coap** is an **OPEN Open Source Project**, see the <a href="#contributing">Contributing</a> section to find out what this means.

<a name="install"></a>
## Installation

```
$ npm install coap --save
```

<a name="basic"></a>
## Basic Example

The following example opens a UDP server and sends a
CoAP message to it:

```js
const coap = require('coap')
const server = coap.createServer()

server.on('request', (req, res) => {
  res.end('Hello ' + req.url.split('/')[1] + '\n')
})

// the default CoAP port is 5683
server.listen(() => {
  const req = coap.request('coap://localhost/Matteo')

  req.on('response', (res) => {
    res.pipe(process.stdout)
    res.on('end', () => {
      process.exit(0)
    })
  })

  req.end()
})
```

or on IPv6:

```js
const coap = require('coap')
const server = coap.createServer({ type: 'udp6' })

server.on('request', (req, res) => {
  res.end('Hello ' + req.url.split('/')[1] + '\n')
})

// the default CoAP port is 5683
server.listen(() => {
  const req = coap.request('coap://[::1]/Matteo')

  req.on('response', (res) => {
    res.pipe(process.stdout)
    res.on('end', () => {
      process.exit(0)
    })
  })

  req.end()
})
```
<a name="proxy"></a>
## Proxy features
The library now comes with the ability to behave as a COAP proxy for other COAP endpoints. In order to activate the
proxy features, create the server with the `proxy` option activated.

A proxy-enabled service behaves as usual for all requests, except for those coming with the `Proxy-Uri` option. This
requests will be redirected to the URL specified in the option, and the response from this option will, in turn,  be
redirected to the caller. In this case, the proxy server handler is not called at all (redirection is automatic).

You can find an example of how this mechanism works in `examples/proxy.js`. This example features one target server
that writes all the information it receives along with the origin port and a proxy server. Once the servers are up:

- Ten requests are sent directly to the server (without reusing ports)
- Ten requests are sent through the proxy (without reusing ports)

The example shows that the target server sees the last ten requests as coming from the same port (the proxy), while the
first ten come from different ports.

<a name="api"></a>
## API

  * <a href="#request"><code>coap.<b>request()</b></code></a>
  * <a href="#createServer"><code>coap.<b>createServer()</b></code></a>
  * <a href="#incoming"><code>IncomingMessage</b></code></a>
  * <a href="#outgoing"><code>OutgoingMessage</b></code></a>
  * <a href="#observeread"><code>ObserveReadStream</b></code></a>
  * <a href="#observewrite"><code>ObserveWriteStream</b></code></a>
  * <a href="#registerOption"><code>coap.<b>registerOption()</b></code></a>
  * <a href="#ignoreOption"><code>coap.<b>ignoreOption()</b></code></a>
  * <a href="#registerFormat"><code>coap.<b>registerFormat()</b></code></a>
  * <a href="#agent"><code>coap.<b>Agent</b></code></a>
  * <a href="#globalAgent"><code>coap.<b>globalAgent</b></code></a>
  * <a href="#globalAgentIPv6"><code>coap.<b>globalAgentIPv6</b></code></a>
  * <a href="#updateTiming"><code>coap.<b>updateTiming</b></code></a>
  * <a href="#defaultTiming"><code>coap.<b>defaultTiming</b></code></a>

-------------------------------------------------------
<a name="request"></a>
### request(requestParams)

Execute a CoAP request. `requestParams` can be a string or an object.
If it is a string, it is parsed using `new URL(requestParams)`.
If it is an object:

- `host`: A domain name or IP address of the server to issue the request
  to.
  Defaults to `'localhost'`.
- `hostname`: To support `URL` `hostname` is preferred over
  `host`
- `port`: Port of remote server. Defaults to 5683.
- `method`: A string specifying the CoAP request method. Defaults to
  `'GET'`.
- `confirmable`: send a CoAP confirmable message (CON), defaults to
  `true`.
- `observe`: send a CoAP observe message, allowing the streaming of
  updates from the server.
- `pathname`: Request path. Defaults to `'/'`. Should not include query string
- `query`: Query string. Defaults to `''`. Should not include the path,
  e.g. 'a=b&c=d'
- `options`: object that includes the CoAP options, for each key-value
  pair the [setOption()](#setOption) will be called.
- `headers`: alias for `options`, but it works only if `options` is
  missing.
- `agent`: Controls [`Agent`](#agent) behavior. Possible values:
  * `undefined` (default): use [`globalAgent`](#globalAgent), a single socket for all
    concurrent requests.
  * [`Agent`](#agent) object: explicitly use the passed in [`Agent`](#agent).
  * `false`: opts out of socket reuse with an [`Agent`](#agent), each request uses a
    new UDP socket.
- `proxyUri`: adds the Proxy-Uri option to the request, so if the request is sent to a
  proxy (or a server with proxy features) the request will be forwarded to the selected URI.
  The expected value is the URI of the target. E.g.: 'coap://192.168.5.13:6793'
- `multicast`: If set to `true`, it forces request to be multicast. Several `response` events
  will be emitted for each received response. It's user's responsibility to set proper multicast `host` parameter
  in request configuration. Default `false`.
- `multicastTimeout`: time to wait for multicast reponses in milliseconds. It is only applicable in case if `multicast` is `true`. Default `20000 ms`.
- `retrySend`: overwrite the default maxRetransmit, useful when you want to use a custom retry count for a request


`coap.request()` returns an instance of <a
href='#outgoing'><code>OutgoingMessage</code></a>.
If you need
to add a payload, just `pipe` into it.
Otherwise, you __must__ call `end` to submit the request.

If `hostname` is a IPv6 address then the payload is sent through a
IPv6 UDP socket, dubbed in node.js as `'udp6'`.

#### Event: 'response'

`function (response) { }`

Emitted when a response is received.
`response` is
an instance of <a
href='#incoming'><code>IncomingMessage</code></a>.

If the `observe` flag is specified, the `'response'` event
will return an instance of
 <a href='#observeread'><code>ObserveReadStream</code></a>.
Which represent the updates coming from the server, according to the
[observe spec](http://tools.ietf.org/html/draft-ietf-core-observe-11).

-------------------------------------------------------
<a name="createServer"></a>
### createServer([options], [requestListener])

Returns a new CoAP Server object.

The `requestListener` is a function which is automatically
added to the `'request'` event.

The constructor can be given an optional options object, containing one of the following options:
* `type`: indicates if the server should create IPv4 connections (`udp4`) or IPv6 connections (`udp6`). Defaults
  to `udp4`.
* `proxy`: indicates that the server should behave like a proxy for incoming requests containing the `Proxy-Uri` header.
  An example of how the proxy feature works, refer to the example in the `/examples` folder. Defaults to `false`.
* `multicastAddress`: Optional. Use this in order to force server to listen on multicast address
* `multicastInterface`: Optional. Use this in order to force server to listen on multicast interface. This is only applicable
  if `multicastAddress` is set. If absent, server will try to listen `multicastAddress` on all available interfaces
* `piggybackReplyMs`: set the number of milliseconds to wait for a
  piggyback response. Default 50.
* `sendAcksForNonConfirmablePackets`: Optional. Use this to suppress sending ACK messages for non-confirmable packages
* `clientIdentifier`: Optional. If specified, it should be a callback function with a signature like `clientIdentifier(request)`, where request is an `IncomingMessage`. The function should return a string that the caches can assume will uniquely identify the client.
* `reuseAddr`: Optional. Use this to specify whether it should be possible to have multiple server instances listening to the same port. Default `true`.

#### Event: 'request'

`function (request, response) { }`

Emitted each time there is a request.
`request` is an instance of <a
href='#incoming'><code>IncomingMessage</code></a> and `response` is
an instance of <a
href='#outgoing'><code>OutgoingMessage</code></a>.

If the `observe` flag is specified, the `response` variable
will return an instance of <a href='#observewrite'><code>ObserveWriteStream</code></a>.
Each `write(data)` to the stream will cause a new observe message sent
to the client.

#### server.listen(port, [address], [callback])

Begin accepting connections on the specified port and hostname.  If the
hostname is omitted, the server will accept connections directed to any
IPv4 or IPv6 address by passing `null` as the address to the underlining socket.

To listen
 to a unix socket, supply a filename instead of port and hostname.

A custom socket object can be passed as a `port` parameter. This custom socket
must be an instance of `EventEmitter` which emits `message`, `error` and
`close` events and implements `send(msg, offset, length, port, address, callback)`
function, just like `dgram.Socket`.
In such case, the custom socket must be pre-configured manually, i.e. CoAP server
will not bind, add multicast groups or do any other configuration.

This function is asynchronous.

#### server.close([callback])

Closes the server.

This function is synchronous, but it provides an asynchronous callback
for convenience.

-------------------------------------------------------
<a name="outgoing"></a>
### OutgoingMessage

An `OutgoingMessage` object is returned by `coap.request` or
emitted by the `coap.createServer` `'response'` event.
It may be used to access response status, headers and data.

It implements the [Writable
Stream](http://nodejs.org/api/stream.html#stream_class_stream_writable) interface, as well as the
following additional properties, methods and events.

#### message.code

The CoAP code of the message.
It is HTTP-compatible, as it can be passed `404`.

#### message.statusCode

(same as message.code)

<a name="setOption"></a>
#### message.setOption(name, value)

Sets a single option value.
All the options are in binary format, except for
`'Content-Format'`, `'Accept'`, `'Max-Age'` and `'ETag'`.
See <a href='#registerOption'><code>registerOption</code></a>
 to know how to register more.

Use an array of buffers
if you need to send multiple options with the same name.

If you need to pass a custom option, pass a string containing a
a number as key and a `Buffer` as value.

Example:

```js
message.setOption('Content-Format', 'application/json')
```

or

```js
message.setOption('555', [Buffer.from('abcde'), Buffer.from('ghi')])
```

`setOption` is also aliased as `setHeader` for HTTP API
compatibility.

Also, `'Content-Type'` is aliased to `'Content-Format'` for HTTP
compatibility.

Since v0.7.0, this library supports blockwise transfers, you can trigger
them by adding a `req.setOption('Block2', Buffer.of(0x2))` to the
output of [request](#request).

And since v0.25.0, this library supports rudimentry type 1 blockwise transfers, you can trigger
them by adding a `req.setOption('Block1', Buffer.of(0x2))` to the
options of [request](#request).

(The hex value 0x2 specifies the size of the blocks to transfer with.
Use values 0 to 6 for 16 to 1024 byte block sizes respectively.)

See the
[spec](http://tools.ietf.org/html/draft-ietf-core-coap-18#section-5.4)
for all the possible options.

#### message.reset()
Returns a Reset COAP Message to the sender. The RST message will appear as an empty message with code `0.00` and the
reset flag set to `true` to the caller. This action ends the interaction with the caller.

#### message.writeHead(code, headers)
Functions somewhat like `http`'s `writeHead()` function.  If `code` is does not match the CoAP code mask of `#.##`, it is coerced into this mask.  `headers` is an object with keys being the header names, and values being the header values.

#### message.on('timeout', function(err) { })
Emitted when the request does not receive a response or acknowledgement within a transaction lifetime.
`Error` object with message `No reply in XXXs` and `retransmitTimeout` property is provided as a parameter.

#### message.on('error', function(err) { })
Emitted when an error occurs. This can be due to socket error, confirmable message timeout or any other generic error.
`Error` object is provided, that describes the error.

-------------------------------------------------------
<a name="incoming"></a>
### IncomingMessage

An `IncomingMessage` object is created by `coap.createServer` or
`coap.request`
and passed as the first argument to the `'request'` and `'response'` event
respectively. It may be used to access response status, headers and data.

It implements the [Readable
Stream](http://nodejs.org/api/stream.html#stream_class_stream_readable) interface, as well as the
following additional methods and properties.

#### message.payload

The full payload of the message, as a Buffer.

#### message.options

All the CoAP options, as parsed by
[CoAP-packet](http://github.com/mcollina/coap-packet).

All the options are in binary format, except for
`'Content-Format'`, `'Accept'` and `'ETag'`.
See <a href='#registerOption'><code>registerOption()</code></a> to know how to register more.

See the
[spec](http://tools.ietf.org/html/draft-ietf-core-coap-18#section-5.4)
for all the possible options.

#### message.headers


CHANGELOG.md
# Changelog

## 0.26.0
*This version supports all current LTS versions of Node.js (12.x, 14.x, 16.x). Older versions might still work but are untested from now on!*

* (JKRhb) Fix coding style with regard to StandardJS
* (JKRhb) chore: update dependencies
* (JKRhb) chore: remove Node version 10.x from CI testing
* (JKRhb) docs: Let content-format registry link point to IANA
* (JKRhb) fix: fix bugs in two examples
* (JKRhb) fix(agent): reset _lastMessageId to 0
* (JKRhb) fix(server): allow limiting to only one server instance per port number
* (JKRhb) feat: add typescript declarations
* (JKRhb) feat: add more Content-Formats
* (JKRhb) test: add more tests for helper functions
* (JKRhb) feat: add support for FETCH, PATCH, and iPATCH
* (JKRhb) chore: add macos-latest to CI pipeline
* (JKRhb) Various refactorings and cleanup
* (JKRhb) Add strict null and type checking
* (JKRhb) Improve README/documentation

## 0.25.0
*This version supports all current LTS versions of Node.js (10.x, 12.x, 14.x, 16.x). Older versions might still work but are untested from now on!*

* (everhardt/JKRhb) feat: add clientIdentifier option to createServer
* (JKRhb) feat: add more content-formats
* (ats-org/Jamezo97/invaderb) feat: Type-1 Block-wise transfer
* (sjlongland/pekkanikander) feat: Add support for de-registering observations
* (phretor) maintenance: let client handle invalid formats themselves instead throwing an error in the library
* (JKRhb) fix: handling of block2 reponses to multicast requests
* (JKRhb) fix: copy all listeners when using block2 multicast
* (JKRhb) fix: three more problems with blockwise multicast responses
* (ats-org/Jamezo97/invaderb) fix: dgram fix
* (JKRhb) fix: several documentation and code (style) optimizations

## 0.24.0
*This version supports all current LTS versions of Node.js (10.x/12.x, 14.x). Older versions might still work but are untested from now on!*

* (blankm) Fix: Retrysend broken when passing socket to agent
* (mateusz-) Fix: fix missing responses to non-confirmable multicast request
* (JcBernack) Fix: update Buffer usage (prevent deprecation wanrings)
* (JsonMa) Fix: fix agent config bug
* (Apollon77) Fix: Add missing checks for Buffer length before reading from it
* (Apollon77) Update all dependencies
* (JsonMa) Fix: remove useless variables
* (JcBernack) Fix: Fix more flag during block-wise transfer

## 0.23.1 (and below)
... can be found in the [Releases list on GitHub](https://github.com/mcollina/node-coap/releases)


lib/agent.ts
/*
 * Copyright (c) 2013-2021 node-coap contributors.
 *
 * node-coap is licensed under an MIT +no-false-attribs license.
 * All rights not explicitly granted in the MIT license are reserved.
 * See the included LICENSE file for more details.
 */

import crypto = require('crypto')
import { Socket, createSocket } from 'dgram'
import { AgentOptions, CoapRequestParams, Block } from '../models/models'
import { EventEmitter } from 'events'
import { parse, generate, ParsedPacket } from 'coap-packet'
import IncomingMessage from './incoming_message'
import OutgoingMessage from './outgoing_message'
import ObserveStream from './observe_read_stream'
import RetrySend from './retry_send'
import { parseBlock2, createBlock2, getOption, removeOption } from './helpers'
import { SegmentedTransmission } from './segmentation'
import { parseBlockOption } from './block'
import { AddressInfo } from 'net'
import { parameters } from './parameters'

const maxToken = Math.pow(2, 32)
const maxMessageId = Math.pow(2, 16)

class Agent extends EventEmitter {
    _opts: AgentOptions
    _closing: boolean
    _sock: Socket | null
    _msgIdToReq: Map<number, OutgoingMessage>
    _tkToReq: Map<string, OutgoingMessage>
    _tkToMulticastResAddr: Map<string, string[]>
    private _lastToken: number
    _lastMessageId: number
    private _msgInFlight: number
    _requests: number
    constructor (opts?: AgentOptions) {
        super()

        if (opts == null) {
            opts = {}
        }

        if (opts.type == null) {
            opts.type = 'udp4'
        }

        if (opts.socket != null) {
            const sock = opts.socket as any
            opts.type = sock.type
            delete opts.port
        }

        this._opts = opts

        this._init(opts.socket)
    }

    _init (socket?: Socket): void {
        this._closing = false

        if (this._sock != null) {
            return
        }

        this._sock = socket ?? createSocket({ type: this._opts.type ?? 'udp4' })
        this._sock.on('message', (msg, rsinfo) => {
            let packet: ParsedPacket
            try {
                packet = parse(msg)
            } catch (err) {
                return
            }

            if (packet.code[0] === '0' && packet.code !== '0.00') {
                // ignore this packet since it's not a response.
                return
            }

            if (this._sock != null) {
                const outSocket = this._sock.address()
                this._handle(packet, rsinfo, outSocket)
            }
        })

        if (this._opts.port != null) {
            this._sock.bind(this._opts.port)
        }

        this._sock.on('error', (err) => {
            this.emit('error', err)
        })

        this._msgIdToReq = new Map()
        this._tkToReq = new Map()
        this._tkToMulticastResAddr = new Map()

        this._lastToken = Math.floor(Math.random() * (maxToken - 1))
        this._lastMessageId = Math.floor(Math.random() * (maxMessageId - 1))

        this._msgInFlight = 0
        this._requests = 0
    }

    _cleanUp (): void {
        if (--this._requests !== 0) {
            return
        }

        if (this._opts.socket == null) {
            this._closing = true
        }

        if (this._msgInFlight !== 0) {
            return
        }

        this._doClose()
    }

    _doClose (): void {
        for (const req of this._msgIdToReq.values()) {
            req.sender.reset()
        }

        if (this._opts.socket != null) {
            return
        }

        if (this._sock != null) {
            this._sock.close()
        }
        this._sock = null
    }

    _handle (packet: ParsedPacket, rsinfo: AddressInfo, outSocket: AddressInfo): void {
        let buf: Buffer
        let response: IncomingMessage
        let req: OutgoingMessage | undefined = this._msgIdToReq.get(packet.messageId)
        const ackSent = (err: Error): void => {
            if (err != null && req != null) {
                req.emit('error', err)
            }

            this._msgInFlight--
            if (this._closing && this._msgInFlight === 0) {
                this._doClose()
            }
        }
        if (req == null) {
            if (packet.token.length > 0) {
                req = this._tkToReq.get(packet.token.toString('hex'))
            }

            if ((packet.ack || packet.reset) && req == null) {
                // Nothing to do on unknown or duplicate ACK/RST packet
                return
            }

            if (req == null) {
                buf = generate({
                    code: '0.00',
                    reset: true,
                    messageId: packet.messageId
                })

                if (this._sock != null) {
                    this._msgInFlight++
                    this._sock.send(buf, 0, buf.length, rsinfo.port, rsinfo.address, ackSent)
                }
                return
            }
        }

        if (packet.confirmable) {
            buf = generate({
                code: '0.00',
                ack: true,
                messageId: packet.messageId
            })

            if (this._sock != null) {
                this._msgInFlight++
                this._sock.send(buf, 0, buf.length, rsinfo.port, rsinfo.address, ackSent)
            }
        }

        if (packet.code !== '0.00' && (req._packet.token == null || req._packet.token.length !== packet.token.length || Buffer.compare(req._packet.token, packet.token) !== 0)) {
        // The tokens don't match, ignore the message since it is a malformed response
            return
        }

        const block1Buff = getOption(packet.options, 'Block1')
        let block1: Block | null = null
        if (block1Buff instanceof Buffer) {
            block1 = parseBlockOption(block1Buff)
            // check for error
            if (block1 == null) {
                req.sender.reset()
                req.emit('error', new Error('Failed to parse block1'))
                return
            }
        }

        req.sender.reset()

        if (block1 != null && packet.ack) {
        // If the client takes too long to respond then the retry sender will send
        // another packet with the previous messageId, which we've already removed.
            const segmentedSender = req.segmentedSender
            if (segmentedSender != null) {
                // If there's more to send/receive, then carry on!
                if (segmentedSender.remaining() > 0) {
                    if (segmentedSender.isCorrectACK(block1)) {
                        if (req._packet.messageId != null) {
                            this._msgIdToReq.delete(req._packet.messageId)
                        }
                        req._packet.messageId = this._nextMessageId()
                        this._msgIdToReq.set(req._packet.messageId, req)
                        segmentedSender.receiveACK(block1)
                    } else {
                        segmentedSender.resendPreviousPacket()
                    }
                    return
                } else {
                    // console.log("Packet received done");
                    if (req._packet.options != null) {
                        removeOption(req._packet.options, 'Block1')
                    }
                    delete req.segmentedSender
                }
            }
        }

        if (!packet.confirmable && !req.multicast) {
            this._msgIdToReq.delete(packet.messageId)
        }

        // Drop empty messages (ACKs), but process RST
        if (packet.code === '0.00' && !packet.reset) {
            return
        }

        const block2Buff = getOption(packet.options, 'Block2')
        let block2: Block | null = null
        // if we got blockwise (2) response
        if (block2Buff instanceof Buffer) {
            block2 = parseBlock2(block2Buff)
            // check for error
            if (block2 == null) {
                req.sender.reset()
                req.emit('error', new Error('failed to parse block2'))
                return
            }
        }
        if (block2 != null) {
            if (req.multicast) {
                req = this._convertMulticastToUnicastRequest(req, rsinfo)
                if (req == null) {
                    return
                }
            }

            // accumulate payload
            req._totalPayload = Buffer.concat([req._totalPayload, packet.payload])

            if (block2.more === 1) {
                // increase message id for next request
                if (req._packet.messageId != null) {
                    this._msgIdToReq.delete(req._packet.messageId)
                }
                req._packet.messageId = this._nextMessageId()
                this._msgIdToReq.set(req._packet.messageId, req)

                // next block2 request
                const block2Val = createBlock2({
                    more: 0,
                    num: block2.num + 1,
                    size: block2.size
                })
                if (block2Val == null) {
                    req.sender.reset()
                    req.emit('error', new Error('failed to create block2'))
                    return
                }
                req.setOption('Block2', block2Val)
                req._packet.payload = undefined
                req.sender.send(generate(req._packet))

                return
            } else {
                // get full payload
                packet.payload = req._totalPayload
                // clear the payload incase of block2
                req._totalPayload = Buffer.alloc(0)
            }
        }

        if (req.response != null) {
            const response: any = req.response
            if (response.append != null) {
                // it is an observe request
                // and we are already streaming
                return response.append(packet)
            } else {
                // TODO There is a previous response but is not an ObserveStream !
                return
            }
        } else if (block2 != null && packet.token != null) {
            this._tkToReq.delete(packet.token.toString('hex'))
        } else if (req.url.observe !== true && !req.multicast) {
            // it is not, so delete the token
            this._tkToReq.delete(packet.token.toString('hex'))
        }

        if (req.url.observe === true && packet.code !== '4.04') {
            response = new ObserveStream(packet, rsinfo, outSocket)
            response.on('close', () => {
                this._tkToReq.delete(packet.token.toString('hex'))
                this._cleanUp()
            })
            response.on('deregister', () => {
                const deregisterUrl = Object.assign({}, req?.url)
                deregisterUrl.observe = 1
                deregisterUrl.token = req?._packet.token

                const deregisterReq = this.request(deregisterUrl)
                // If the request fails, we'll deal with it with a RST message anyway.
                deregisterReq.on('error', () => {})
                deregisterReq.end()
            })
        } else {
            response = new IncomingMessage(packet, rsinfo, outSocket)
        }

        if (!req.multicast) {
            req.response = response
        }

        req.emit('response', response)
    }

    _nextToken (): Buffer {
        const buf = Buffer.alloc(8)

        if (++this._lastToken === maxToken) {
            this._lastToken = 0
        }

        buf.writeUInt32BE(this._lastToken, 0)
        crypto.randomBytes(4).copy(buf, 4)

        return buf
    }

    _nextMessageId (): number {
        if (++this._lastMessageId === maxMessageId) {
            this._lastMessageId = 0
        }

        return this._lastMessageId
    }

    /**
     * Entry point for a new client-side request.
     * @param url The parameters for the request
     */
    request (url: CoapRequestParams): OutgoingMessage {
        this._init()

        const options = url.options ?? url.headers
        const multicastTimeout = url.multicastTimeout != null ? url.multicastTimeout : 20000
        const host = url.hostname ?? url.host
        const port = url.port ?? parameters.coapPort

        const req = new OutgoingMessage({}, (req, packet) => {
            let buf

            if (url.confirmable !== false) {
                packet.confirmable = true
            }

            // multicast message should be forced non-confirmable
            if (url.multicast === true) {
                req.multicast = true
                packet.confirmable = false
            }

            if (!(packet.ack ?? packet.reset ?? false)) {
                packet.messageId = this._nextMessageId()
                if ((url.token instanceof Buffer) && (url.token.length > 0)) {
                    if (url.token.length > 8) {
                        return req.emit('error', new Error('Token may be no longer than 8 bytes.'))
                    }
                    packet.token = url.token
                } else {
                    packet.token = this._nextToken()
                }
                const token = packet.token.toString('hex')
                if (req.multicast) {
                    this._tkToMulticastResAddr.set(token, [])
                }
                if (token != null) {
                    this._tkToReq.set(token, req)
                }
            }

            if (packet.messageId != null) {
                this._msgIdToReq.set(packet.messageId, req)
            }

            const block1Buff = getOption(packet.options, 'Block1')
            if (block1Buff != null) {
                // Setup for a segmented transmission
                req.segmentedSender = new SegmentedTransmission(block1Buff[0], req, packet)
                req.segmentedSender.sendNext()
            } else {
                try {
                    buf = generate(packet)
                } catch (err) {
                    req.sender.reset()
                    return req.emit('error', err)
                }
                req.sender.send(buf, packet.confirmable === false)
            }
        })

        req.sender = new RetrySend(this._sock, port, host, url.retrySend)

        req.url = url

        req.statusCode = url.method ?? 'GET'

        this.urlPropertyToPacketOption(url, req, 'pathname', 'Uri-Path', '/')
        this.urlPropertyToPacketOption(url, req, 'query', 'Uri-Query', '&')

        if (options != null) {
            for (const optionName of Object.keys(options)) {
                if (optionName in options) {
                    req.setOption(optionName, options[optionName])
                }
            }
        }

        if (url.proxyUri != null) {
            req.setOption('Proxy-Uri', url.proxyUri)
        }

        req.sender.on('error', req.emit.bind(req, 'error'))

        req.sender.on('sending', () => {
            this._msgInFlight++
        })

        req.sender.on('timeout', (err) => {
            req.emit('timeout', err)
            this.abort(req)
        })

        req.sender.on('sent', () => {
            if (req.multicast) {
                return
            }

            this._msgInFlight--
            if (this._closing && this._msgInFlight === 0) {
                this._doClose()
            }
        })

        // Start multicast monitoring timer in case of multicast request
        if (url.multicast === true) {
            req.multicastTimer = setTimeout(() => {
                if (req._packet.token != null) {
                    const token = req._packet.token.toString('hex')
                    this._tkToReq.delete(token)
                    this._tkToMulticastResAddr.delete(token)
                }
                if (req._packet.messageId != null) {
                    this._msgIdToReq.delete(req._packet.messageId)
                }
                this._msgInFlight--
                if (this._msgInFlight === 0 && this._closing) {
                    this._doClose()
                }
            }, multicastTimeout)
        }

        if (typeof (url.observe) === 'number') {
            req.setOption('Observe', url.observe)
        } else if (url.observe == null) {
            req.on('response', this._cleanUp.bind(this))
        } else if (url.observe) {
            req.setOption('Observe', 0)
        }

        this._requests++

        req._totalPayload = Buffer.alloc(0)

        return req
    }

    abort (req: OutgoingMessage): void {
        req.sender.removeAllListeners()
        req.sender.reset()
        this._cleanUp()
        if (req._packet.messageId != null) {
            this._msgIdToReq.delete(req._packet.messageId)
        }
        if (req._packet.token != null) {
            this._tkToReq.delete(req._packet.token.toString('hex'))
        }
    }

    urlPropertyToPacketOption (url: CoapRequestParams, req: OutgoingMessage, property: string, option: string, separator: string): void {
        if (url[property] != null) {
            req.setOption(option, url[property].normalize('NFC').split(separator)
                .filter((part) => { return part !== '' })
                .map((part) => {
                    const buf = Buffer.alloc(Buffer.byteLength(part))
                    buf.write(part)
                    return buf
                }))
        }
    }

    _convertMulticastToUnicastRequest (req: any, rsinfo: AddressInfo): OutgoingMessage | undefined {
        const unicastReq = this.request(req.url)
        const unicastAddress = rsinfo.address.split('%')[0]
        const token = req._packet.token.toString('hex')
        const addressArray = this._tkToMulticastResAddr.get(token) ?? []

        if (addressArray.includes(unicastAddress)) {
            return undefined
        }

        unicastReq.url.host = unicastAddress
        unicastReq.sender._host = unicastAddress
        clearTimeout(unicastReq.multicastTimer)
        unicastReq.url.multicast = false
        req.eventNames().forEach(eventName => {
            req.listeners(eventName).forEach(listener => {
                unicastReq.on(eventName, listener)
            })
        })
        addressArray.push(unicastAddress)
        unicastReq._packet.token = this._nextToken()
        this._requests++
        return unicastReq
    }
}

export default Agent
